ebi_eva_common_pyutils/nextflow/nextflow_pipeline.py

```diff
@@ -0,0 +1,181 @@
+# Copyright 2021 EMBL - European Bioinformatics Institute
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Rationale for a Nextflow pipeline abstraction
+# ---------------------------------------------
+# Dynamic pipeline generation
+# Abstraction to represent process dependencies
+# Unit testability of individual steps without scattering logic between Python and Nextflow
+# Ability to combine pipelines
```

Comment:
Just my two cents on this - I think this abstraction is cool, it's clear and easy to understand on its own, and as it's already proved useful I'm happy with it. :+1:

I think there's some cognitive overhead to having it around, as it's a different way of thinking about Nextflow pipelines where you explicitly model the dependencies rather than having them implicit in the flow of data (and I think some of the other benefits might be covered by [DSL 2](https://www.nextflow.io/docs/latest/dsl2.html) which we're already using elsewhere...).  But I don't necessarily think that's a problem, if it helps us get more things done more effectively.

Reply:
Fair point! I did extensive research on DSL 2 before embarking on this. I agree that DSL 2 offers ways to model dependencies (through channels and functions) + an awkward way to combine pipelines through modules: so it sort of ticks one-and-a-half box above :). But it offers no support for unit-testing inputs/outputs of individual steps, control flow or integration testing. All of that aside, my primary impetus for this module was dynamic pipeline generation (useful for distributed processing similar to EVA-2389) which this abstraction offers a good support for.